### ISSUE TITLE
no-unused-prop-types - cannot detect usage within componentWillReceiveProps

### ISSUE THREAD

**Issue by ctumolosus** [yannickcr/eslint-plugin-react#884]
I have a component where I'm only referencing a prop within `componentWillReceiveProps(nextProps)` and `react/no-unused-prop-types` is not picking it up.
It's probably a bad practice that I need to revisit on my end, but wondering if you guys think it's something the rule should handle.

**Comment by ljharb** [comment]
Can you provide the exact code in question?

**Comment by ctumolosus** [comment]
Below is a simplified example. We have a custom higher-order component that handles form submission, validation, etc. and it injects a `hasSubmitted` property into the wrapped component when submitted successfully. The form is not editable until the edit button is pressed. It's trivial to refactor this example to satisfy the rule, but I'm curious if the rule should handle this scenario.

```jsx
import React, { Component, PropTypes } from 'react';

export default class UserForm extends Component {

  static propTypes = {
    firstName: PropTypes.string,
    lastName: PropTypes.string,
    onSubmit: PropTypes.func,
    hasSubmitted: PropTypes.bool,
  }

  static defaultProps = {
    hasSubmitted: false,
  }

  state = {
    isEditable: false,
  };

  componentWillReceiveProps(nextProps) {
    if (nextProps.hasSubmitted) {
      this.setState({ isEditable: false });
    }
  }

  handleEdit = () => {
    this.setState({ isEditable: true });
  }

  render() {
    return (
      <form onSubmit={this.props.onSubmit}>
        <label htmlFor="firstName">First Name</label>
        <input type="text" id="firstName" name="firstName" value={this.props.firstName} disabled={!this.state.isEditable} />
        <label htmlFor="lastName">Last Name</label>
        <input type="text" id="lastName" name="lastName" value={this.props.lastName} disabled={!this.state.isEditable} />
        {this.state.isEditable ? (
          <div>
            <button>Cancel</button>
            <button type="submit">Update</button>
          </div>
        ) : (
          <div>
            <button onTouchTap={this.handleEdit}>Edit</button>
          </div>
        )}
      </form>
    );
  }
}
```

**Comment by EvNaverniouk** [comment]
I'm unable to reproduce the issue using the code your provided. The rule currently supports prop checking for `componentWillReceiveProps`.

Can you confirm you're running the latest version of `eslint-plugin-react` (6.3.0)

**Comment by ctumolosus** [comment]
Here is repository replicating issue: https://github.com/ctumolosus/eslint-plugin-react-issue-884

If you run `npm run lint` after installing all dependencies you should see the error. Please let me know if there is anything else I can do to help.

**Comment by marcgrr** [comment]
I think I have encountered this same bug. I noticed that the bug only appears if I access the prop with a dot (`nextProps.hasSubmitted`). If I access the prop with destructuring (`const { hasSubmitted } = nextProps`), then `react/no-unused-prop-types` correctly recognizes the access.

I'm going to see if I can make a PR fixing this!

**Comment by imontiel** [comment]
It looks like you have to destructure the props from `nextProps` – and it **has** to be `nextProps`. It can't be any other name.

**Comment by DianaSuvorova** [comment]
This is fixed.
@ljharb  I'd be happy do it myself if you give me tights to close tickets.

Thanks :)

**Comment by ljharb** [comment]
That's not up to me :-) thanks for all your triage help!

**Comment by nischalashashidhar** [comment]
@marcgrr , Thanks.. It worked for me as well.
